File: android/app/src/main/java/ae/findmybay/data/api/AuthApi.kt
package ae.findmybay.data.api

import ae.findmybay.data.api.dto.AuthResponse
import ae.findmybay.data.api.dto.GoogleSignInBody
import ae.findmybay.data.api.dto.MeDto
import ae.findmybay.data.api.dto.OtpRequestBody
import ae.findmybay.data.api.dto.OtpRequestResponse
import ae.findmybay.data.api.dto.OtpVerifyBody
import ae.findmybay.data.api.dto.UpdateMeBody
import ae.findmybay.data.api.dto.UpdateProfileBody
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Headers
import retrofit2.http.PATCH
import retrofit2.http.POST

interface AuthApi {

    /** Request an OTP for the given UAE phone number. No auth required. */
    @Headers("No-Auth: true")
    @POST("auth/otp/request")
    suspend fun requestOtp(@Body body: OtpRequestBody): OtpRequestResponse

    /** Verify the OTP and get back JWT tokens. No auth required. */
    @Headers("No-Auth: true")
    @POST("auth/otp/verify")
    suspend fun verifyOtp(@Body body: OtpVerifyBody): AuthResponse

    /**
     * Exchange a Google ID token (from GoogleSignInClient) for our JWT pair.
     * The backend verifies the token against Google's JWKs and links/creates
     * the user by sub + email. No auth required.
     */
    @Headers("No-Auth: true")
    @POST("auth/google")
    suspend fun signInWithGoogle(@Body body: GoogleSignInBody): AuthResponse

    /** Returns the current user's profile. */
    @GET("auth/me")
    suspend fun me(): MeDto

    /** Patches editable profile fields. */
    @PATCH("auth/me")
    suspend fun updateMe(@Body body: UpdateMeBody): MeDto

    /**
     * Saves the customer's car profile (mobile + car details + plate).
     * Idempotent + partial — the onboarding flow posts each step separately
     * and the same endpoint is used later from a "Profile" settings page.
     */
    @PATCH("auth/me/profile")
    suspend fun updateCarProfile(@Body body: UpdateProfileBody): MeDto
}

File: ios/FindMyBay/Data/API/AuthAPI.swift
// AuthAPI.swift
//
// Auth endpoints — parity with `data/api/AuthApi.kt` (Retrofit interface).

import Foundation

protocol AuthAPI {
    func requestOtp(_ body: OtpRequestBody) async throws -> OtpRequestResponse
    func verifyOtp(_ body: OtpVerifyBody) async throws -> AuthResponse
    func signInWithGoogle(_ body: GoogleSignInBody) async throws -> AuthResponse
    func signInWithApple(_ body: AppleSignInBody) async throws -> AuthResponse
    func refresh(_ body: RefreshBody) async throws -> RefreshResponse
    func me() async throws -> MeDTO
    func updateMe(_ body: UpdateMeBody) async throws -> MeDTO
    func updateCarProfile(_ body: UpdateCarProfileBody) async throws -> MeDTO
}

struct AuthAPIImpl: AuthAPI {
    let client: APIClient

    func requestOtp(_ body: OtpRequestBody) async throws -> OtpRequestResponse {
        try await client.request(
            .POST, path: "auth/otp/request", body: body,
            headers: [kNoAuthHeader: "true"]
        )
    }

    func verifyOtp(_ body: OtpVerifyBody) async throws -> AuthResponse {
        try await client.request(
            .POST, path: "auth/otp/verify", body: body,
            headers: [kNoAuthHeader: "true"]
        )
    }

    func signInWithGoogle(_ body: GoogleSignInBody) async throws -> AuthResponse {
        try await client.request(
            .POST, path: "auth/google", body: body,
            headers: [kNoAuthHeader: "true"]
        )
    }

    func signInWithApple(_ body: AppleSignInBody) async throws -> AuthResponse {
        try await client.request(
            .POST, path: "auth/apple", body: body,
            headers: [kNoAuthHeader: "true"]
        )
    }

    func refresh(_ body: RefreshBody) async throws -> RefreshResponse {
        // No-Auth: we're explicitly trading the refresh token for a new pair,
        // and any stale access token in the keychain would leak into the
        // header otherwise.
        try await client.request(
            .POST, path: "auth/refresh", body: body,
            headers: [kNoAuthHeader: "true"]
        )
    }

    func me() async throws -> MeDTO {
        try await client.request(.GET, path: "auth/me")
    }

    func updateMe(_ body: UpdateMeBody) async throws -> MeDTO {
        try await client.request(.PATCH, path: "auth/me", body: body)
    }

    func updateCarProfile(_ body: UpdateCarProfileBody) async throws -> MeDTO {
        try await client.request(.PATCH, path: "auth/me/profile", body: body)
    }
}
